smelli/unbinned.py

```diff
@@ -0,0 +1,78 @@
+from flavio.io import instanceio as iio
+import numpy as np
+import yaml
+
+class UnbinnedParameterLikelihood(iio.YAMLLoadable):
+    """An `UnbinnedParameterLikelihood` provides an unbinned likelihood function in terms of
+    parameters.
+    Methods:
+    - `log_likelihood`: The likelihood as a function of the parameters
+    Instances can be imported and exported from/to YAML using the `load`
+    and `dump` methods.
+    """
+    _input_schema_dict = {
+        'eft': str,
+        'basis': str,
+        'scale': float,
+        'names': list,
+        'samples': np.array,
+        'weights': np.array
+    }
+
+    _output_schema_dict = {
+        'eft': str,
+        'basis': str,
+        'scale': float,
+        'names': list,
+        'samples': np.array,
+        'weights': np.array
+    }
+
+    def __init__(self,
+                 eft,
+                 basis,
+                 scale,
+                 names,
+                 samples,
+                 weights=None):
+        """Initialize the instance.
+        Parameters:
+        - `eft`: a string identifying the WET as defined by wcxf
+        - `basis`: a string identifying the WET basis as defined by wcxf
+        - `scale`: a scale (in units of GeV) at which the WET coefficient samples are valid
+        - `names`: an ordered list of names identiying the Wilson coefficients in the WET basis
+        - `samples`: a numpy array of vectors of Wilson coefficients, with each row representing a sample
+        - `weights`: a numpy array of scalar weights on a logarithmic scale for each sample
+        """
+        self.eft     = eft
+        self.basis   = basis
+        self.scale   = scale
+        self.names   = names
+        self.samples = samples
+        self.weights = weights if weights is None else np.ones(len(samples))
+
+    def log_likelihood(self, smeft_wc):
+        """Return the likelihood for all parameters."""
+
+        wet_wc = smeft_wc.match_run(self.scale, self.eft, self.basis)
+        central = [wet_wc[name] for name in self.names]
+        result = 0.0
+        for weight, sample in zip(self.weights, self.samples):
+            result += weight * np.sum(np.absolute(sample - central)**2)
+
+        return result
+
+    @classmethod
+    def load(cls, f, **kwargs):
+        """Instantiate an object from a YAML string or stream."""
+        d = yaml.load(f, Loader=yaml.Loader)
```

Comment:
@DavidMStraub @peterstangl I'm overriding the use of the ```SafeLoader``` here since I can dump but not load the ```np.array``` otherwise. Any recommendations?

Reply:
Are you storing an actual numpy array in a YAML file? If it's just an array of numbers, I would suggest using the `tolist()` method and store it as a native YAML (nested) list. This would avoid such problems.

Then again, there is a question of design: is it really a good idea to store all the samples in the YAML? I mean, in practice, this will be a huge file, so wouldn't it make more sense to store it somewhere more convenient (up to the user, `.npy`, HDF5, ...) and use the YAML only for metadata?

Reply:
@DavidMStraub's advice has been followed. Using .npy files.